docs/fundamentals/code-analysis/quality-rules/ca1863.md

```diff
@@ -0,0 +1,94 @@
+---
+title: "CA1863: Use 'CompositeFormat'"
+description: "Learn about code analysis rule CA1863: Use 'CompositeFormat'"
+ms.date: 11/20/2023
+f1_keywords:
+  - "CA1863"
+helpviewer_keywords:
+  - "CA1863"
+dev_langs:
+  - CSharp
+  - VB
+---
+
+# CA1863: Use 'CompositeFormat'
+
+| Property                            | Value                                  |
+|-------------------------------------|----------------------------------------|
+| **Rule ID**                         | CA1863                                 |
+| **Title**                           | Use `CompositeFormat`                  |
+| **Category**                        | [Performance](performance-warnings.md) |
+| **Fix is breaking or non-breaking** | Non-breaking                           |
+| **Enabled by default in .NET 8**    | No                                     |
+
+## Cause
+
+Code calls <xref:System.String.Format(System.String,System.Object[])?displayProperty=nameWithType> or <xref:System.Text.StringBuilder.AppendFormat(System.String,System.Object[])?displayProperty=nameWithType> with a `static` format string that hasn't been cached to a <xref:System.Text.CompositeFormat> instance.
+
+## Rule description
+
+It's expensive to parse a format string at run time. This rule locates places in your code where you can cache and use a <xref:System.Text.CompositeFormat> instance as the argument to a formatting operation, rather than passing in the original format string. A <xref:System.Text.CompositeFormat> instance parses the composite format string when it's created, which means the "hot path" of string formatting can execute much faster.
+
+## How to fix violations
+
+Create an instance of <xref:System.Text.CompositeFormat> by calling <xref:System.Text.CompositeFormat.Parse(System.String)?displayProperty=nameWithType> and pass it to <xref:M:System.String.Format(System.IFormatProvider,System.Text.CompositeFormat,System.Object[])?displayProperty=nameWithType> or <xref:System.Text.StringBuilder.AppendFormat(System.IFormatProvider,System.Text.CompositeFormat,System.Object[])?displayProperty=nameWithType> instead of the original format string.
+
+## Example
+
+The following example shows two violations of the rule:
+
+```csharp
+class C
+{
+    private static readonly string StaticField;
+
+    static void Main()
+    {
+        _ = string.Format(StaticField, 42);
```

Comment:
@stephentoub I was unable to get the rule to fire on the equivalent VB code. Here's what I tried:

```vb
Class C
    Private Shared ReadOnly s1 As String
    Sub M1()

        Dim unused = String.Format(s1, 42)

        Dim sb As New StringBuilder()
        sb.AppendFormat(s1, 42)
    End Sub
End Class
```

Reply:
It's quite possible it doesn't work for VB currently. It was only tested for C#.